File: source_py/tests/test_scanner.py
import os
import tempfile
import time
import unittest
from pathlib import Path

from ebook_renamer.scanner import Scanner


class TestScanner(unittest.TestCase):
    def setUp(self):
        self.test_dir = tempfile.TemporaryDirectory()
        self.root_path = self.test_dir.name

    def tearDown(self):
        self.test_dir.cleanup()

    def create_file(self, filename, content="content"):
        path = os.path.join(self.root_path, filename)
        with open(path, "w") as f:
            f.write(content)
        return path

    def test_scanner_creates_correct_file_info(self):
        # Create content larger than 1KB
        large_content = "This is a test file that is definitely larger than 1KB. " * 50
        self.create_file("test_book.pdf", large_content)

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        file_info = files[0]
        self.assertEqual(file_info.original_name, "test_book.pdf")
        self.assertEqual(file_info.extension, ".pdf")
        self.assertFalse(file_info.is_failed_download)
        self.assertFalse(file_info.is_too_small)

    def test_scanner_detects_tar_gz(self):
        self.create_file("arXiv-2012.08669v1.tar.gz")

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertEqual(files[0].extension, ".tar.gz")

    def test_scanner_detects_download_files(self):
        self.create_file("test_book.pdf.download", "")

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertTrue(files[0].is_failed_download)

    def test_scanner_detects_small_files(self):
        self.create_file("tiny.pdf", "x")  # 1 byte

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertTrue(files[0].is_too_small)

    def test_scanner_skips_hidden_files(self):
        self.create_file(".hidden.pdf")

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 0)

    def test_scanner_skips_download_directories(self):
        download_dir = os.path.join(self.root_path, "some_book.download")
        os.mkdir(download_dir)
        with open(os.path.join(download_dir, "content.pdf"), "w") as f:
            f.write("content")

        scanner = Scanner(self.root_path, 2)
        files = scanner.scan()

        self.assertEqual(len(files), 0)

    # ========== EDGE CASE TESTS ==========

    def test_scanner_detects_crdownload_files(self):
        """Chrome partial downloads use .crdownload extension"""
        self.create_file("test_book.pdf.crdownload", "partial content")

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertTrue(files[0].is_failed_download)
        self.assertEqual(files[0].extension, ".crdownload")

    def test_scanner_empty_directory(self):
        """Empty directory should return no files"""
        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 0)

    def test_scanner_multiple_file_types(self):
        large_content = "x" * 2000

        self.create_file("book.pdf", large_content)
        self.create_file("book.epub", large_content)
        self.create_file("notes.txt", large_content)
        self.create_file("archive.tar.gz", large_content)
        self.create_file("document.mobi", large_content)

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 5)

        extensions = [f.extension for f in files]
        self.assertIn(".pdf", extensions)
        self.assertIn(".epub", extensions)
        self.assertIn(".txt", extensions)
        self.assertIn(".tar.gz", extensions)
        self.assertIn(".mobi", extensions)

    def test_scanner_nested_directories(self):
        """Test that scanner can find files in nested directories"""
        large_content = "x" * 2000

        # Create file at depth 1
        self.create_file("book1.pdf", large_content)

        # Create file at depth 2
        subdir = os.path.join(self.root_path, "subdir")
        os.mkdir(subdir)
        with open(os.path.join(subdir, "book2.pdf"), "w") as f:
            f.write(large_content)

        # Create file at depth 3
        subsubdir = os.path.join(subdir, "subsubdir")
        os.mkdir(subsubdir)
        with open(os.path.join(subsubdir, "book3.pdf"), "w") as f:
            f.write(large_content)

        # With sufficient depth, should find all files
        scanner = Scanner(self.root_path, 10)
        files = scanner.scan()
        self.assertEqual(len(files), 3)

    def test_scanner_epub_small_file_detection(self):
        self.create_file("tiny.epub", "x")

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertTrue(files[0].is_too_small)

    def test_scanner_size_threshold_exactly_1kb(self):
        # Exactly 1024 bytes should NOT be too small
        self.create_file("exact_1kb.pdf", "x" * 1024)

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertFalse(files[0].is_too_small)

    def test_scanner_size_threshold_1023_bytes(self):
        # 1023 bytes should be too small
        self.create_file("under_1kb.pdf", "x" * 1023)

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertTrue(files[0].is_too_small)

    def test_scanner_file_without_extension(self):
        self.create_file("README", "content")

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertEqual(files[0].extension, "")
        self.assertEqual(files[0].original_name, "README")

    def test_scanner_unicode_filename(self):
        large_content = "x" * 2000
        self.create_file("数学入門.pdf", large_content)

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertEqual(files[0].original_name, "数学入門.pdf")
        self.assertEqual(files[0].extension, ".pdf")

    def test_scanner_file_size_recorded(self):
        content = "Hello, World!"  # 13 bytes
        self.create_file("small.txt", content)

        scanner = Scanner(self.root_path, 1)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertEqual(files[0].size, 13)

    def test_scanner_preserves_relative_path_structure(self):
        large_content = "x" * 2000
        subdir = os.path.join(self.root_path, "books", "2024")
        os.makedirs(subdir)
        with open(os.path.join(subdir, "book.pdf"), "w") as f:
            f.write(large_content)

        scanner = Scanner(self.root_path, 10)
        files = scanner.scan()

        self.assertEqual(len(files), 1)
        self.assertIn("books", str(files[0].original_path))
        self.assertIn("2024", str(files[0].original_path))

File: source_rb/test/test_scanner.rb
require 'test_helper'
require 'minitest/autorun'
require 'fileutils'
require 'tempfile'

class TestScanner < Minitest::Test
  def setup
    @test_dir = Dir.mktmpdir
    @scanner = Scanner.new(@test_dir, 5)
  end

  def teardown
    FileUtils.rm_rf(@test_dir)
  end

  def test_scan_pdf_files
    File.write(File.join(@test_dir, "test.pdf"), "%PDF-1.4 content")
    File.write(File.join(@test_dir, "book.epub"), "epub content")
    File.write(File.join(@test_dir, "notes.txt"), "text content")
    
    files = @scanner.scan
    
    assert_equal 3, files.length
    assert_includes files.map { |f| f[:original_name] }, "test.pdf"
    assert_includes files.map { |f| f[:original_name] }, "book.epub"
    assert_includes files.map { |f| f[:original_name] }, "notes.txt"
  end

  def test_skip_hidden_files
    File.write(File.join(@test_dir, ".hidden.pdf"), "hidden content")
    File.write(File.join(@test_dir, "visible.pdf"), "visible content")
    
    files = @scanner.scan
    
    assert_equal 1, files.length
    assert_includes files.map { |f| f[:original_name] }, "visible.pdf"
    refute_includes files.map { |f| f[:original_name] }, ".hidden.pdf"
  end

  def test_max_depth_limit
    FileUtils.mkdir_p(File.join(@test_dir, "level1"))
    FileUtils.mkdir_p(File.join(@test_dir, "level1", "level2"))
    FileUtils.mkdir_p(File.join(@test_dir, "level1", "level2", "level3"))
    
    File.write(File.join(@test_dir, "root.pdf"), "root level")
    File.write(File.join(@test_dir, "level1", "file1.pdf"), "level 1")
    File.write(File.join(@test_dir, "level1", "level2", "file2.pdf"), "level 2")
    File.write(File.join(@test_dir, "level1", "level2", "level3", "file3.pdf"), "level 3")
    
    # Test with max depth 2
    scanner = Scanner.new(@test_dir, 2)
    files = scanner.scan
    
    assert_equal 3, files.length
    assert_includes files.map { |f| f[:original_name] }, "root.pdf"
    assert_includes files.map { |f| f[:original_name] }, "file1.pdf"
    assert_includes files.map { |f| f[:original_name] }, "file2.pdf"
    refute_includes files.map { |f| f[:original_name] }, "file3.pdf"
  end

  def test_detect_file_extensions
    File.write(File.join(@test_dir, "book.pdf"), "pdf content")
    File.write(File.join(@test_dir, "document.epub"), "epub content")
    File.write(File.join(@test_dir, "notes.txt"), "text content")
    File.write(File.join(@test_dir, "archive.tar.gz"), "archive content")
    File.write(File.join(@test_dir, "incomplete.download"), "download content")
    
    files = @scanner.scan
    
    assert_equal 6, files.length
    
    file_extensions = files.map { |f| f[:extension] }
    assert_includes file_extensions, ".pdf"
    assert_includes file_extensions, ".epub"
    assert_includes file_extensions, ".txt"
    assert_includes file_extensions, ".tar.gz"
    assert_includes file_extensions, ".download"
  end

  def test_file_size_detection
    # Large PDF (should not be too small)
    large_content = "x" * 2000  # ~2KB
    File.write(File.join(@test_dir, "large.pdf"), large_content)
    
    # Small PDF (should be too small)
    File.write(File.join(@test_dir, "small.pdf"), "x")  # 1 byte
    
    # TXT file (should not be checked for size)
    File.write(File.join(@test_dir, "notes.txt"), "x")
    
    files = @scanner.scan
    
    pdf_files = files.select { |f| f[:extension] == ".pdf" }
    small_files = pdf_files.select { |f| f[:is_too_small] }
    txt_files = files.select { |f| f[:extension] == ".txt" }
    
    assert_equal 2, pdf_files.length
    assert_equal 1, small_files.length
    assert_equal 1, small_files.first[:original_name], "small.pdf"
    assert_equal 0, txt_files.select { |f| f[:is_too_small] }.length
  end

  def test_failed_download_detection
    File.write(File.join(@test_dir, "book.pdf"), "valid pdf")
    File.write(File.join(@test_dir, "incomplete.download"), "partial download")
    File.write(File.join(@test_dir, "stuck.crdownload"), "chrome download")
    
    files = @scanner.scan
    
    download_files = files.select { |f| f[:is_failed_download] }
    
    assert_equal 2, download_files.length
    download_names = download_files.map { |f| f[:original_name] }
    assert_includes download_names, "incomplete.download"
    assert_includes download_names, "stuck.crdownload"
  end

  def test_nested_directory_scanning
    FileUtils.mkdir_p(File.join(@test_dir, "subdir1", "subdir2"))
    File.write(File.join(@test_dir, "subdir1", "file1.pdf"), "content 1")
    File.write(File.join(@test_dir, "subdir2", "file2.pdf"), "content 2")
    
    files = @scanner.scan
    
    assert_equal 2, files.length
    assert_includes files.map { |f| f[:original_name] }, "file1.pdf"
    assert_includes files.map { |f| f[:original_name] }, "file2.pdf"
  end

  def test_invalid_path_handling
    # Test that scanner handles various path issues gracefully
    assert_raises(StandardError) do
      Scanner.new("/nonexistent/path/that/does/not/exist", 1)
    end
    
    assert_raises(StandardError) do
      Scanner.new(nil, 1)
    end
  end

  def test_unicode_filenames
    # Test with Unicode filenames
    File.write(File.join(@test_dir, "数学入門.pdf"), "unicode content")
    File.write(File.join(@test_dir, "éléphant.pdf"), "accented content")
    
    files = @scanner.scan
    
    assert_equal 2, files.length
    unicode_names = files.map { |f| f[:original_name] }
    assert_includes unicode_names, "数学入門.pdf"
    assert_includes unicode_names, "éléphant.pdf"
  end

  def test_symlink_handling
    # Create a symlink and test if scanner handles it safely
    real_file = File.join(@test_dir, "real.pdf")
    File.write(real_file, "real content")
    
    symlink_file = File.join(@test_dir, "symlink.pdf")
    begin
      File.symlink(real_file, symlink_file)
    rescue NotImplementedError
      skip "Symlinks not supported on this platform"
    end
    
    files = @scanner.scan
    
    # Should either skip or resolve symlinks to real files
    if File.symlink?(symlink_file)
      # This test ensures we don't count symlinks twice
      real_files = files.select { |f| f[:original_name] == "real.pdf" }
      symlink_files = files.select { |f| f[:original_name] == "symlink.pdf" }
      
      assert_equal 1, real_files.length
      assert_operator :<=, symlink_files.length, 1
    end
  end

  def test_file_info_structure
    File.write(File.join(@test_dir, "test.pdf"), "pdf content")
    
    files = @scanner.scan
    file_info = files.first
    
    # Ensure file info has expected structure
    assert_respond_to :original_name, file_info
    assert_respond_to :extension, file_info
    assert_respond_to :size, file_info
    assert_respond_to :original_path, file_info
    assert_respond_to :is_too_small, file_info
    assert_respond_to :is_failed_download, file_info
  end

  def test_empty_directory
    # Test scanning empty directory
    files = @scanner.scan
    
    assert_equal [], files
  end

  def test_permissions_error_handling
    # Test that scanner handles permission errors gracefully
    begin
      # Create a directory we can't read
      restricted_dir = File.join(@test_dir, "restricted")
      FileUtils.mkdir_p(restricted_dir)
      
      # Remove read permissions
      FileUtils.chmod(0000, restricted_dir)
      
      files = @scanner.scan
      
      # Should skip or handle the restricted directory
      assert files.all? { |f| !f[:original_path].include?(restricted_dir) }
    ensure
      # Restore permissions for cleanup
      FileUtils.chmod(0755, restricted_dir)
    end
  end

  private

  def assert_respond_to(method, file_info)
    assert_respond_to method, file_info, "Expected to respond to #{method}"
  end
end